File: fourierDivision.py
def split(n, base):
    t = []
    extra = False
    
    while (n > 0):
        t.insert(0, n % base)
        n = int(n / base)

    if (len(t) % 2 == 1):
        t.append(0)
        extra = True

    a = []
    for i in range(0, len(t), 2):
        a.append(t[i] * base + t[i + 1])
    if (len(a) == 1):
        a.append(0)

    return a, extra

def comb(l, base):
    n = 0
    for e in l:
        n = n*(base**2) + e
    return n

def div(dividend, divisor, base, trunc, round = False):
    c, cExt = split(dividend, base)
    a, aExt = split(divisor, base)

    factor = 1
    if aExt:
        factor *= base
    if cExt:
        factor /= base
    
    b = []
    tmp = (c[0] * (base**2) + c[1])
    b.append(int(tmp / a[0]))
    r = tmp % a[0]

    for i in range(2, len(c) - len(a) + 1):
        tmp = r * (base**2) + c[i]
        for j in range(1, min(len(a), i)):
            tmp -= a[j] * b[i - j - 1]
        b.append(int(tmp / a[0]))
        r = tmp % a[0]

    res = comb(b, base) * factor
    if (len(a) >= len(c)):
        res /= (base**2)**(1 + len(a) - len(c))

    if (trunc):
        return int(res) + (1 if round and r >= (base / 2) else 0)

    return res

def test(dividend, divisor, base = 10, trunc = True):
    dq = div(dividend, divisor, base, trunc)
    q = dividend/divisor
    if (trunc):
        q = int(q)

    print(str(dividend) + " / " + str(divisor) + " = " + str(q))
    
    if (dq == q):
        print("pass")
    else:
        print("fail: " + str(dq) + " diff : " + str(abs(dq - q)))

    print()

bases = [10, 16, 32768]
for base in bases:
    print("Testing with base: " + str(base) + "\n")
    
    test(499084777422, base, 6534)
    test(100000, 314159, base, False)
    test(875352, 6543, base)
    test(876342615243, 73524377, base)
    test(654, 23, base)
    test(87897432432, 67676237, base)
    test(6633554627152354545, 856352525354556, base)
    test(100000000000, 314159265359, base, False)
    test(12345678, 123, base)
    test(12345678, 12345, base)
    test(12345678, 1234567, base)

File: fourierDivision.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <utility>

#include <inttypes.h>
#include <math.h>
#include <string.h>

using namespace std;

#define MAX_ERROR 0.0000001

#ifdef SMALL
#define BASE 10
#define BASE_SQR 100
#else
#define BASE 65536
#define BASE_SQR 4294967296ULL
#endif


template <typename VecObject>
void printList(const VecObject &v) {
   #ifndef DEBUG
   return;
   #endif

   typename VecObject::const_iterator it;
   for (it = v.begin(); it != v.end(); ++it)
      cout << *it << " ";
   cout << endl;
}

pair<vector<uint32_t>, int> split(uint64_t number) {
   vector<uint32_t> t;
   int extra = 0;

   while (number > 0) {
      t.insert(t.begin(), number % BASE);
      number /= BASE;
   }

   if (t.size() % 2 == 1) {
      t.push_back(0);
      extra++;
   }

   vector<uint32_t> a;
   for (int i = 0; i < t.size(); i += 2)
      a.push_back(t[i] * BASE + t[i + 1]);

   if (a.size() == 1) {
      a.push_back(0);
      extra += 2;
   }

   return make_pair(a, extra);
}

uint64_t comb(vector<int64_t> l) {
   uint64_t n = 0;
   for (vector<int64_t>::iterator it = l.begin(); it != l.end(); it++)
      n = n*BASE_SQR + *it;
   return n;
}

void div(uint64_t dividend, uint64_t divisor, void *res, bool trunc) {
   if (divisor == 0 || dividend == 0) {
      memset(res, (divisor == 0 ? 0xFF : 0), 8);
      return;
   }

   pair<vector<uint32_t>, int> t;

   t = split(divisor);
   vector<unsigned int> a = t.first;
   int aExt = t.second;
   printList(a);

   t = split(dividend);
   vector<unsigned int> c = t.first;
   int cExt = t.second;
   printList(c);

   vector<int64_t> b;
   int64_t tmp = c[0] * BASE_SQR + c[1];
   b.push_back(tmp / a[0]);
   int64_t r = tmp % a[0];

   int limit = c.size() - a.size() + 1;
   for (int i = 2; i < limit; i++) {
      tmp = r * BASE_SQR + c[i];

      for (int j = 1; j < a.size() && j < i; j++)
         tmp -= a[j] * b[i - j - 1];

      b.push_back(tmp / a[0]);
      r = tmp % a[0];
   }

   printList(b);

   double scale = pow(BASE, (aExt - cExt) - 2 * max(0, (int) (1 + a.size() - c.size())));
   if (trunc)
      *((uint64_t *) res) = comb(b) * scale;
   else
      *((double *) res) = comb(b) * scale;

   // *((double *) res) = ((double) comb(b)) * pow(BASE, aExt - cExt) / pow(BASE_SQR, 1 + a.size() - c.size());
}

void test(uint64_t dividend, uint64_t divisor, bool trunc = true) {
   double fd, d;
   uint64_t fi, i;

   if (divisor == 0) {
      cout << "Divide by 0" << endl;
      div(dividend, divisor, &fi, trunc);
      cout << "handled " << (~fi == 0 ? "correctly" : "incorrectly") << endl << endl;
      return;
   }

   if (trunc) {
      div(dividend, divisor, &fi, trunc);
      i = dividend / divisor;
   } else {
      div(dividend, divisor, &fd, trunc);
      d = ((double) dividend) / divisor;
   }

   double err = trunc ? (fi > i ? (fi - i) : (i - fi)) : fabs(fd - d);

   cout << dividend << " / " << divisor << " = " << (trunc ? i : d) << endl;
   cout << (err < MAX_ERROR ? "pass" : "fail") << " error: " << err << endl << endl;
}

int main() {
   test(499084777422ULL, 6534);
   test(100000, 314159, false);
   test(875352, 6543);
   test(876342615243ULL, 73524377);
   test(654, 23);
   test(87897432432ULL, 67676237);
   test(6633554627152354545ULL, 856352525354556ULL);
   test(100000000000ULL, 314159265359ULL, false);
   test(12345678, 123);
   test(12345678, 12345);
   test(12345678, 1234567);
   test(0, 65567536576355ULL);
   test(0, 67868764857673ULL, false);
   test(767565, 0);
   test(7657785, 0, false);
   test(854848695528570185ULL, 1454337869);
   test(854848695528570185ULL, 587792365);
   return 0;
}